plugins/modem-manager/fu-mbim-qdu-updater.c

```diff
@@ -0,0 +1,464 @@
+		mbim_device_command (ctx->mbim_device,
+				     request,
+				     20,
+				     ctx->cancellable,
+				     (GAsyncReadyCallback)fu_mbim_qdu_updater_file_write_ready,
+				     ctx);
+		mbim_message_unref (request);
+	} else {
+		fu_device_set_progress (FU_DEVICE (ctx->device), 100);
+		g_debug ("Done!");
+		fu_device_set_status (FU_DEVICE (ctx->device), FWUPD_STATUS_DEVICE_RESTART);
+		g_warn_if_fail (ctx->timeout_id == 0);
+		/* device will auto reboot right after update finish */
+		ctx->timeout_id = g_timeout_add_seconds (30, fu_mbim_qdu_updater_reboot_timeout, ctx);
+	}
+}
+
+static void
+fu_mbim_qdu_updater_file_open_ready (MbimDevice *device, GAsyncResult *res, gpointer user_data)
+{
+	g_autoptr(MbimMessage) request = NULL;
+	g_autoptr(MbimMessage) response = NULL;
+	g_autoptr(GError)      error = NULL;
+	WriteContext *ctx = user_data;
+	guint32  out_max_transfer_size;
+	guint32  out_max_window_size;
+	guint32  total_sending_numbers = 0;
+	FuChunk *chk = NULL;
+
+	response = mbim_device_command_finish (device, res, &error);
+	if (!response || !mbim_message_response_get_result (response, MBIM_MESSAGE_TYPE_COMMAND_DONE, &error)) {
+		g_debug ("error: operation failed: %s\n", error->message);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (!mbim_message_qdu_file_open_response_parse (response,
+							&out_max_transfer_size,
+							&out_max_window_size,
+							&error)) {
+		g_debug ("error: couldn't parse response message: %s\n", error->message);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (response)
+		mbim_message_unref (response);
+
+	total_sending_numbers = (g_bytes_get_size (ctx->blob)) / (out_max_transfer_size);
+	if ((g_bytes_get_size (ctx->blob)) % (out_max_transfer_size) != 0)
+		total_sending_numbers ++;
+
+	ctx->chunks = fu_chunk_array_new_from_bytes (ctx->blob,
+						     0x00,	/* start addr */
+						     0x00,	/* page_sz */
+						     out_max_transfer_size);
+	chk = g_ptr_array_index (ctx->chunks, 0);
+	request = mbim_message_qdu_file_write_set_new (fu_chunk_get_data_sz (chk), (const guint8 *)fu_chunk_get_data (chk), NULL);
+	mbim_device_command (ctx->mbim_device,
+			     request,
+			     10,
+			     ctx->cancellable,
+			     (GAsyncReadyCallback)fu_mbim_qdu_updater_file_write_ready,
+			     ctx);
+	mbim_message_unref (request);
+}
+
+static void
+fu_mbim_qdu_updater_session_ready (MbimDevice *device, GAsyncResult *res, gpointer user_data)
+{
+	g_autoptr(MbimMessage) response = NULL;
+	MbimQduSessionType     out_current_session_type;
+	MbimQduSessionStatus   out_current_session_status;
+	MbimQduSessionType     out_last_session_type;
+	MbimQduSessionResult   out_last_session_result;
+	guint32                out_last_session_error_offset;
+	guint32                out_last_session_error_size;
+	g_autoptr(GError)      error = NULL;
+	g_autoptr(MbimMessage) request = NULL;
+	WriteContext *ctx = user_data;
```

Comment:
This last variable seems to be unaligned w.r.t. the others? maybe align it  as well.